src/main/java/com/es/phoneshop/service/HttpSessionCartService.java

```diff
@@ -0,0 +1,70 @@
+package com.es.phoneshop.service;
+
+import com.es.phoneshop.dao.ArrayListProductDao;
+import com.es.phoneshop.dao.ProductDao;
+import com.es.phoneshop.exception.NullValuePassedException;
+import com.es.phoneshop.exception.OutOfStockException;
+import com.es.phoneshop.model.cart.Cart;
+import com.es.phoneshop.model.cart.CartItem;
+import com.es.phoneshop.model.product.Product;
+
+import javax.servlet.http.HttpServletRequest;
+import java.util.Optional;
+
+public class HttpSessionCartService implements CartService {
+    private static volatile CartService instance;
+    private static final String CART_SESSION_ATTRIBUTE = HttpSessionCartService.class.getName() + ".cart";
+    private ProductDao productDao;
+
+    private HttpSessionCartService() {
+        productDao = ArrayListProductDao.getInstance();
+    }
+
+    public static CartService getInstance() {
+        if (instance == null) {
+            synchronized (HttpSessionCartService.class) {
+                if (instance == null) {
+                    instance = new HttpSessionCartService();
+                }
+            }
+        }
+        return instance;
+    }
+
+    @Override
+    public synchronized Cart getCart(HttpServletRequest request) {
+        Cart cart = (Cart) request.getSession().getAttribute(CART_SESSION_ATTRIBUTE);
+        if (cart == null) {
+            request.getSession().setAttribute(CART_SESSION_ATTRIBUTE, cart = new Cart());
+        }
+        return cart;
+    }
+
+    @Override
+    public synchronized void add(Cart cart, Long productId, int quantity) throws OutOfStockException {
+        try {
+            Product product = productDao.getProduct(productId);
+            Optional<CartItem> optionalCartItem = cart.getItems().stream()
+                    .filter(cartItem -> productId.equals(cartItem.getProduct().getId()))
+                    .findFirst();
+            if (optionalCartItem.isPresent()) {
+                int totalQuantity = quantity + optionalCartItem.get().getQuantity();
+                if (product.getStock() < totalQuantity) {
+                    throw new OutOfStockException(product, quantity, product.getStock());
+                }
+                optionalCartItem.get().setQuantity(totalQuantity);
```

Comment:
User should not be able to have negative amount of products in cart. Though he might need to add negative amount to remove some if he'd like.